File: packages/typescript/src/resources/signals.ts
import type { HttpClient } from '../http.js';
import type { Signal, SignalQuery, CreateSignalBody, BulkAcknowledgeSignalsBody, SignalStats } from '../types/signal.js';

export class SignalsResource {
  constructor(private http: HttpClient) {}

  async list(opts?: SignalQuery): Promise<{ signals: Signal[]; next_cursor: string | null }> {
    return this.http.get<{ signals: Signal[]; next_cursor: string | null }>('/v1/signals', {
      params: opts as Record<string, string | number | boolean | undefined>,
    });
  }

  async get(id: string): Promise<Signal> {
    return this.http.get<Signal>(`/v1/signals/${id}`);
  }

  async create(body: CreateSignalBody): Promise<Signal> {
    return this.http.post<Signal>('/v1/signals', body);
  }

  async acknowledge(id: string): Promise<Signal> {
    return this.http.patch<Signal>(`/v1/signals/${id}/acknowledge`, {});
  }

  async acknowledgeBulk(body: BulkAcknowledgeSignalsBody): Promise<{ acknowledged: number }> {
    return this.http.post<{ acknowledged: number }>('/v1/signals/acknowledge-bulk', body);
  }

  async stats(): Promise<SignalStats> {
    return this.http.get<SignalStats>('/v1/signals/stats');
  }
}

File: packages/python/invariance/resources/signals.py
from __future__ import annotations

from typing import Any

from ..http_client import HttpClient


class SignalsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self, opts: dict[str, Any] | None = None) -> dict[str, Any]:
        return await self._http.get("/v1/signals", params=opts)

    async def get(self, signal_id: str) -> dict[str, Any]:
        return await self._http.get(f"/v1/signals/{signal_id}")

    async def evidence(self, signal_id: str) -> dict[str, Any]:
        return await self._http.get(f"/v1/signals/{signal_id}/evidence")

    async def create(self, body: dict[str, Any]) -> dict[str, Any]:
        return await self._http.post("/v1/signals", body)

    async def acknowledge(self, signal_id: str) -> dict[str, Any]:
        return await self._http.patch(f"/v1/signals/{signal_id}/acknowledge", {})

    async def acknowledge_bulk(self, body: dict[str, Any]) -> dict[str, Any]:
        return await self._http.post("/v1/signals/acknowledge-bulk", body)

    async def stats(self) -> dict[str, Any]:
        return await self._http.get("/v1/signals/stats")
